File: apps/api/app/schemas/backtesting.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

ResolvedOutcome = Literal["yes", "no", "cancelled"]


class MarketOutcomeCreate(BaseModel):
    resolved_outcome: ResolvedOutcome
    resolved_at: datetime | None = None
    source: str | None = None
    notes: str | None = Field(default=None, max_length=4000)


class MarketOutcomeUpdate(BaseModel):
    resolved_outcome: ResolvedOutcome | None = None
    resolved_at: datetime | None = None
    source: str | None = None
    notes: str | None = Field(default=None, max_length=4000)


class MarketOutcomeRead(BaseModel):
    market_id: int
    question: str
    resolved_outcome: ResolvedOutcome
    resolved_at: datetime
    source: str
    notes: str | None = None


class MarketOutcomesResponse(BaseModel):
    total_count: int = 0
    items: list[MarketOutcomeRead] = Field(default_factory=list)


class BacktestingFamilySummary(BaseModel):
    prediction_family: str
    total_resolved_with_predictions: int = 0
    correct_direction_count: int = 0
    accuracy_direction: Decimal | None = None
    avg_confidence: Decimal | None = None
    brier_score: Decimal | None = None


class BacktestingSummaryResponse(BaseModel):
    generated_at: datetime
    total_resolved_with_predictions: int = 0
    correct_direction_count: int = 0
    accuracy_direction: Decimal | None = None
    avg_confidence: Decimal | None = None
    brier_score: Decimal | None = None
    by_prediction_family: list[BacktestingFamilySummary] = Field(default_factory=list)

File: apps/web/app/lib/backtesting.ts
"use client";

import { fetchApiJson, isApiNotFoundError } from "./api";

export type ResolvedOutcome = "yes" | "no" | "cancelled" | "invalid" | "unknown";

export type MarketOutcome = {
  market_id: number;
  question: string;
  resolved_outcome: ResolvedOutcome;
  resolved_at: string;
  source: string;
  notes?: string | null;
};

export type MarketOutcomePayload = {
  resolved_outcome: ResolvedOutcome;
  resolved_at?: string | null;
  source?: string | null;
  notes?: string | null;
};

export type MarketOutcomesResponse = {
  total_count: number;
  items: MarketOutcome[];
};

export type BacktestingFamilySummary = {
  prediction_family: string;
  total_resolved_with_predictions: number;
  correct_direction_count: number;
  accuracy_direction?: string | number | null;
  avg_confidence?: string | number | null;
  brier_score?: string | number | null;
};

export type BacktestingConfidenceBucket = {
  bucket: string;
  min_confidence: string | number;
  max_confidence: string | number;
  total_resolved_with_predictions: number;
  correct_direction_count: number;
  accuracy_direction?: string | number | null;
  avg_confidence?: string | number | null;
  brier_score?: string | number | null;
};

export type BacktestingSummary = {
  generated_at: string;
  total_outcomes: number;
  total_predictions: number;
  resolved_with_predictions: number;
  total_resolved_with_predictions: number;
  correct_direction_count: number;
  accuracy_direction?: string | number | null;
  avg_confidence?: string | number | null;
  brier_score?: string | number | null;
  by_prediction_family: BacktestingFamilySummary[];
  by_confidence_bucket: BacktestingConfidenceBucket[];
};

async function requestBacktesting<T>(path: string, init?: RequestInit): Promise<T> {
  return fetchApiJson<T>(path, init);
}

export function fetchBacktestingSummary(): Promise<BacktestingSummary> {
  return requestBacktesting<BacktestingSummary>("/backtesting/summary");
}

export function fetchMarketOutcomes(): Promise<MarketOutcomesResponse> {
  return requestBacktesting<MarketOutcomesResponse>("/outcomes?limit=100");
}

export async function fetchMarketOutcome(
  marketId: number | string,
): Promise<MarketOutcome | null> {
  try {
    return await requestBacktesting<MarketOutcome>(`/markets/${marketId}/outcome`);
  } catch (error) {
    if (isApiNotFoundError(error)) {
      return null;
    }
    throw error;
  }
}

export function upsertMarketOutcome(
  marketId: number | string,
  payload: MarketOutcomePayload,
): Promise<MarketOutcome> {
  return requestBacktesting<MarketOutcome>(`/markets/${marketId}/outcome`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateMarketOutcome(
  marketId: number | string,
  payload: Partial<MarketOutcomePayload>,
): Promise<MarketOutcome> {
  return requestBacktesting<MarketOutcome>(`/markets/${marketId}/outcome`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteMarketOutcome(marketId: number | string): Promise<null> {
  return requestBacktesting<null>(`/markets/${marketId}/outcome`, {
    method: "DELETE",
  });
}
